### PR TITLE
Some more lanzatool refactoring

### PR DESCRIPTION
**Pull request by nikstur**
Implements some requested changes from #34.

### CODE REVIEW

**Comment by RaitoBezarius** [comment]
@nikstur Do you think we could get the "alternative Nix store" path feature and make it a parameter or something smart?

File: rust/lanzatool/src/pe.rs [hunk 1]
```diff
@@ -167,17 +167,21 @@ fn write_to_tmp(
...
-        .expect("Failed to convert path '{}' to a relative string path")
-        .replace('/', "\\");
-    Ok(format!("\\{}", &relative_path_string))
+        .with_context(|| format!("Failed to strip esp prefix: {:?} from: {:?}", esp, path))?;
+    let uefi_path = uefi_path(relative_path)?;
+    Ok(format!("\\{}", &uefi_path))
+}
+
+/// Convert a path to a valid UEFI string representation.
+fn uefi_path(path: &Path) -> Result<String> {
```
Comment (RaitoBezarius) [inline rust/lanzatool/src/pe.rs:180]: IMHO, a UEFI String is a UTF-16 string, but String is a UTF-8 String AFAIK.

Comment (RaitoBezarius) [reply rust/lanzatool/src/pe.rs:180]: This function is more a "UNIX paths to Windows-style paths" AFAIK

Comment (RaitoBezarius) [reply rust/lanzatool/src/pe.rs:180]: I wonder if `windows::Path` could be directly imported.

**Comment by nikstur** [comment]
> @nikstur Do you think we could get the "alternative Nix store" path feature and make it a parameter or something smart?

From some research, I got the impression, that an alternative Nix Store path is not _really_ a well-supported feature of Nix e.g. https://github.com/NixOS/nix/issues/1971. I'd say we just do not explicitly support an alternative path. With the implementation in this PR, lanzatool doesn't really know about /nix/store at all anyways, so other store locations should just work

File: rust/lanzatool/src/pe.rs (hunk 1 above)
Comment (nikstur) [reply rust/lanzatool/src/pe.rs:180]: From what I can tell you can't easily convert a path using the stdlib. I found this library: https://github.com/r-efi/r-efi-string.
From the readme it seems that UEFI strings are even more complicated than just UTF-16. Do you think the comment or function name is not fitting? I'm not sure we gain much if we encode the complexities of UEFI Strings into e.g. the function name.

**Comment by RaitoBezarius** [comment]
> > @nikstur Do you think we could get the "alternative Nix store" path feature and make it a parameter or something smart?
> 
> From some research, I got the impression, that an alternative Nix Store path is not _really_ a well-supported feature of Nix e.g. [NixOS/nix#1971](https://github.com/NixOS/nix/issues/1971). 

It definitely is, the issue you linked is about rewriting the paths which is something different and could be useful to reuse the NixOS cache in an alternative store path.
But, if you do `nix-build x --store ~/nix`, it will create a new Nix store at this path.

Some people do run Nix store paths at alternative locations, certainly, this is very edgy, but I'm not sure if it cost that much to support it as a flag.

> I'd say we just do not explicitly support an alternative path. With the implementation from #42 lanzatool doesn't really know about /nix/store at all anyways, so other store locations should just work. 

Hm, that's fair. We can wait for a feature request then.

> I implemented it that way so I can test with files in a temporary directory instead of /nix/store. (See: https://github.com/nix-community/lanzaboote/blob/master/rust/lanzatool/src/esp.rs#L56)

Great!

File: rust/lanzatool/src/pe.rs (hunk 1 above)
Comment (RaitoBezarius) [reply rust/lanzatool/src/pe.rs:180]: In that case, I'd just add a warning that it's not exactly a UEFI-spec compliant function, and it's fine IMHO.

**Review by blitz — APPROVED** [review]
Love the unit tests!

**Pull request merged**